src/companion.jl

```diff
@@ -1,100 +1,108 @@
 export Companion
+using LinearAlgebra: dot
 
 """
-[`Companion` matrix](http://en.wikipedia.org/wiki/Companion_matrix)
+   Companion(v::Union{AbstractVector,Polynomial})::AbstractMatrix
 
-```julia
-julia> A=Companion([3,2,1])
-3x3 Companion{Int64}:
+Construct a lazy
+[`companion` matrix](http://en.wikipedia.org/wiki/Companion_matrix)
+from the coefficients of its characteristic (monic) polynomial.
+
+The matrix is `n × n` for a vector input of length `n`
+or an input `Polynomial` of degree `n`,
+but the storage here is only `O(n)`.
+This version puts the coefficients
+along the last column of the matrix.
+Some texts put the coefficients along the first row,
+the transpose of the convention used here.
+
+This type has efficient methods
+for `mul!` and `inv`.
+
+```jldoctest
+julia> A = Companion([3,2,1])
+3×3 Companion{Int64}:
  0  0  -3
  1  0  -2
  0  1  -1
 ```
 Also, directly from a polynomial:
 
-```julia
+```jldoctest
 julia> using Polynomials
 
-julia> P=Polynomial([2.0,3,4,5])
-Polynomial(2 + 3x + 4x^2 + 5x^3)
+julia> P = Polynomial(2:5)
+Polynomials.Polynomial(2 + 3*x + 4*x^2 + 5*x^3)
 
-julia> C=Companion(P)
+julia> C = Companion(P)
 3×3 Companion{Float64}:
  0.0  0.0  -0.4
  1.0  0.0  -0.6
  0.0  1.0  -0.8
 ```
 """
-struct Companion{T} <: AbstractArray{T, 2}
+struct Companion{T} <: AbstractMatrix{T}
     c :: Vector{T}
 end
 
-# Generate companion matrix from a polynomial
+Companion(v::AbstractVector) = Companion(collect(v))
```

Comment:
Rather than forcing a `collect`, I recommend doing the following. This will use the `convert` mechanism.

```suggestion
Companion(v::AbstractVector{T}) where T = Companion{T}(v)
```